mitmproxy/utils/protoc.py

```diff
@@ -0,0 +1,112 @@
+            message = self.message_factory.GetPrototype(grpc_method.input_type)()
+        elif isinstance(http_message, mitmproxy.http.Response):
+            message = self.message_factory.GetPrototype(grpc_method.output_type)()
+        else:
+            raise ValueError(f"Unexpected HTTP message type {http_message}")
+
+        message.Clear()
+
+        try:
+            message.MergeFromString(data_without_prefix)
+        except DecodeError as e:
+            raise ValueError("Unable to deserialize input") from e
+
+        return MessageToString(
+            message=message,
+            descriptor_pool=self.descriptor_pool,
+            print_unknown_fields=True,
+        )
+
+    def serialize(self, http_message: mitmproxy.http.Message, path: str, text: str) -> bytes:
+        """
+        Takes a string and serializes it into a protobuf byte array.
+        You must set a descriptor file must prior to calling this method.
+        The string must be formatted according `google.protobuf.text_format`
+
+        Raises:
+            ValueError - in case serialization fails because the method could not be resolved or the input data is invalid
+                         e.g. unknown fields present or invalid text format.
+        """
+
+        grpc_method = self.__find_method_by_path(path)
+
+        if isinstance(http_message, mitmproxy.http.Request):
+            empty_message = self.message_factory.GetPrototype(grpc_method.input_type)()
+        elif isinstance(http_message, mitmproxy.http.Response):
+            empty_message = self.message_factory.GetPrototype(grpc_method.output_type)()
+        else:
+            raise ValueError(f"Unexpected HTTP message type {http_message}")
+
+        empty_message.Clear()
+
+        try:
+            populated_message = Parse(
+                text=text,
+                message=empty_message,
+                allow_field_number=True,
+                allow_unknown_field=False,
+                descriptor_pool=self.descriptor_pool
+            )
+        except ParseError as e:
+            raise ValueError("Unable to serialize input") from e
+
+        serializedMessage = populated_message.SerializeToString(deterministic=True)
+        # Prepend the length and compression header; 5 bytes in total in big endian order.
+        # Payload compression is not supported at the moment, so compression bit is always 0.
+        return len(serializedMessage).to_bytes(5, "big") + serializedMessage
+
+    def __find_method_by_path(self, path: str) -> MethodDescriptor:
+        try:
+            # Drop the first '/' from the path and convert the rest to a fully qualified namespace that we can look up.
+            method_path = path.replace("/", ".")[1:]
+            return self.descriptor_pool.FindMethodByName(method_path)
+        except KeyError as e:
+            raise ValueError("Failed to resolve method name by path") from e
+
+
+serializer = ProtocSerializer()
```

Comment:
`ProtocSerializer` is used in both addons and the contentview, so I need to instantiate it earlier. I don't work with python on a daily basis, so I'm unsure if the approach I took is correct. Feel free to propose something better.